bindings/python/setup.py

```diff
@@ -448,10 +497,43 @@ def _configure_b2_with_distutils(self) -> Iterator[None]:
                     log.info(config.read())
                 yield
             finally:
-                config_path.unlink()
+                with contextlib.suppress(FileNotFoundError):
+                    config_path.unlink()
         else:
             yield
 
+    def _find_project_config(self) -> Optional[pathlib.Path]:
+        for directory in itertools.chain(
+            (PYTHON_BINDING_DIR,), PYTHON_BINDING_DIR.parents
+        ):
+            path = directory / "project-config.jam"
+            if path.exists():
+                return path
+        return None
+
+    def _configure_from_autotools(self) -> None:
```

Comment:
did you intend to keep the autotools logic here? It was removed in `RC_2_0`

Reply:
I say leave it in. Then `setup.py` is identical across branches and any changes can be merged everywhere. There's not much downside to leaving it, since it spams a deprecation warning and doesn't have any effect in `RC_2_0` anyway.

Reply:
The cost of leaving it in is maintenance cost of recent branches, the value of leaving it in is maintenance cost in older branches.

It sounds like you anticipate more maintenance of setup.py in RC_1_2 than I do

Reply:
The driving force of the complexity of this file is the "quirks" in boost.build that vary across version, platform and distro. These apply to all branches.

AFAICT boost.build's design is highly ad-hoc so I see no way to be confident we've covered all relevant features, or all aspects of any given one. Git history shows I've iterated a lot on each feature.

So I anticipate endless maintenance of `setup.py` in `RC_1_2`, as long as boost.build is our primary build system.

Today with the branches out of sync, I have to figure out how a fix should apply in each branch, test each, and communicate to you how the `RC_1_2` PR should look when merged to later branches. Each of those steps is time-consuming and brittle. It's discouraging.

IMO branch-specific code is a small cost to gain easier maintenance across all branches.

I propose we follow the pattern I established here: branch-specific features get isolated in their own function, which is enabled by `setup.cfg` in its branch.

Reply:
> So I anticipate endless maintenance of setup.py in RC_1_2, as long as boost.build is our primary build system.

I don't intend to maintain `RC_1_2` indefinitely. (just like 1.1 and earlier aren't maintained)

> IMO branch-specific code is a small cost to gain easier maintenance across all branches.
> 
> I propose we follow the pattern I established here: branch-specific features get isolated in their own function, which is enabled by setup.cfg in its branch.

If the branch specific functions are separated out, it should be fairly easy to merge even if the outdated functionality is gone in future branches. I would expect any conflicts to be easily resolved at least.

I'm not excited about having autoconf logic remain in 2.0+, It's misleading. If you think it's unreasonable to only keep it in RC_1_2, would it be reasonable to remove it alltogether, from RC_1_2 also?

Reply:
> I don't intend to maintain `RC_1_2` indefinitely. (just like 1.1 and earlier aren't maintained)

The point is that most build infra changes should go on all active branches, so should start on `RC_1_2` and be merged through
 
> If the branch specific functions are separated out, it should be fairly easy to merge even if the outdated functionality is gone in future branches. I would expect any conflicts to be easily resolved at least.

Merge conflicts are a manual process. So far, they've been prone to creating spurious differences between branches.

We could audit the branches so they only contain expected changes between them, but that's a manual process.

The discipline of maintaining branch-specific code in isolated functions is a manual process. As soon as someone does a PR against `setup.py` without this discipline and you merge it, the discipline is lost.

Gaps in any of the manual processes create more spurious differences across branches, which create more merge conflicts, which make changes more expensive.

*Maybe* all the manual processes can be followed perfectly until `RC_1_2` is deprecated and autotools logic can be deleted. But it's hard to convince myself that's true when it hasn't been true so far.

Keeping `setup.py` identical on all branches allows for automatic merges in place of these manual processes.

I agree that intentionally keeping dead code is surprising, but it just seems like a small cost to pay, compared to how costly it is to do build system changes today. (#7073 was written 3 weeks ago).

Maintaining three active branches of such an enormous codebase is hard. I don't think it's crazy to pay some costs to make it easier.

All that said, if you insist on removing autotools logic from `RC_2_0`, I'll make that change. But only if you insist :)

> I'm not excited about having autoconf logic remain in 2.0+, It's misleading. If you think it's unreasonable to only keep it in RC_1_2, would it be reasonable to remove it alltogether, from RC_1_2 also?

I do think we should drop autotools from `setup.py`. I did that when I first started working on `setup.py`, it led to #5763

IMO, breaking changes to the build system in a point release is a defensible decision, but you're the one who would have to do the defending.